# EDA - Mall Customer

In [133]:
import numpy as np 

### Load Data Using Numpy genfromtxt

In [134]:
data = np.genfromtxt('Mall_Customers.csv', dtype=['i8','U6','i8','i8','i8'], names=['id', 'Gender', 'Age', 'AnnualIncome','Spending'], delimiter=',', skip_header=1)
data

array([(  1, 'Male', 19,  15, 39), (  2, 'Male', 21,  15, 81),
       (  3, 'Female', 20,  16,  6), (  4, 'Female', 23,  16, 77),
       (  5, 'Female', 31,  17, 40), (  6, 'Female', 22,  17, 76),
       (  7, 'Female', 35,  18,  6), (  8, 'Female', 23,  18, 94),
       (  9, 'Male', 64,  19,  3), ( 10, 'Female', 30,  19, 72),
       ( 11, 'Male', 67,  19, 14), ( 12, 'Female', 35,  19, 99),
       ( 13, 'Female', 58,  20, 15), ( 14, 'Female', 24,  20, 77),
       ( 15, 'Male', 37,  20, 13), ( 16, 'Male', 22,  20, 79),
       ( 17, 'Female', 35,  21, 35), ( 18, 'Male', 20,  21, 66),
       ( 19, 'Male', 52,  23, 29), ( 20, 'Female', 35,  23, 98),
       ( 21, 'Male', 35,  24, 35), ( 22, 'Male', 25,  24, 73),
       ( 23, 'Female', 46,  25,  5), ( 24, 'Male', 31,  25, 73),
       ( 25, 'Female', 54,  28, 14), ( 26, 'Male', 29,  28, 82),
       ( 27, 'Female', 45,  28, 32), ( 28, 'Male', 35,  28, 61),
       ( 29, 'Female', 40,  29, 31), ( 30, 'Female', 23,  29, 87),
       ( 31, 'Male', 

**Note:**

- **Use U6 instead of S6**

**Reference**
- https://numpy.org/doc/stable/reference/generated/numpy.genfromtxt.html
- https://stackoverflow.com/questions/33655641/b-character-added-when-using-numpy-loadtxt

In [135]:
#Check data type of array
type(data)

numpy.ndarray

In [136]:
# Check data type of individual columns
data.dtype

dtype([('id', '<i8'), ('Gender', '<U6'), ('Age', '<i8'), ('AnnualIncome', '<i8'), ('Spending', '<i8')])

In [137]:
# Check shape of array
data.shape

(200,)

In [138]:
# Check dimension of array
data.ndim

1

In [139]:
# Extract first record
first_record = data[0]
first_record

(1, 'Male', 19, 15, 39)

In [140]:
# Check data type of first record
type(first_record)

numpy.void

It looks like tuple but it actually is numpy.void data type.

- https://stackoverflow.com/questions/25247190/how-to-use-numpy-void-type

In [141]:
# Filter gender with male
# Will use the list to double check when doing conversion
data['Gender'] == 'Male'

array([ True,  True, False, False, False, False, False, False,  True,
       False,  True, False, False, False,  True,  True, False,  True,
        True, False,  True,  True, False,  True, False,  True, False,
        True, False, False,  True, False,  True,  True, False, False,
       False, False, False, False, False,  True,  True, False, False,
       False, False, False, False, False, False,  True, False,  True,
       False,  True, False,  True, False,  True,  True,  True, False,
       False,  True,  True, False, False,  True, False,  True, False,
       False, False,  True,  True, False,  True, False, False,  True,
        True,  True, False, False,  True, False, False, False, False,
       False,  True,  True, False, False,  True, False, False,  True,
        True, False, False,  True,  True,  True, False, False,  True,
        True,  True,  True, False, False,  True, False, False, False,
       False, False, False,  True, False, False,  True, False, False,
        True,  True,

**We will be converting Male to 1 and Female to 0.**

In [142]:
# Try np.where without actual assignment. Check if conversion is ok.
np.where(data['Gender'] == 'Male', 1, 0)

array([1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1,
       0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1,
       0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0,
       0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1,
       0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0,
       0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0,
       1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1,
       1, 1])

In [143]:
# Perform actual conversion
data['Gender'] = np.where(data['Gender'] == 'Male', 1, 0)
data

array([(  1, '1', 19,  15, 39), (  2, '1', 21,  15, 81),
       (  3, '0', 20,  16,  6), (  4, '0', 23,  16, 77),
       (  5, '0', 31,  17, 40), (  6, '0', 22,  17, 76),
       (  7, '0', 35,  18,  6), (  8, '0', 23,  18, 94),
       (  9, '1', 64,  19,  3), ( 10, '0', 30,  19, 72),
       ( 11, '1', 67,  19, 14), ( 12, '0', 35,  19, 99),
       ( 13, '0', 58,  20, 15), ( 14, '0', 24,  20, 77),
       ( 15, '1', 37,  20, 13), ( 16, '1', 22,  20, 79),
       ( 17, '0', 35,  21, 35), ( 18, '1', 20,  21, 66),
       ( 19, '1', 52,  23, 29), ( 20, '0', 35,  23, 98),
       ( 21, '1', 35,  24, 35), ( 22, '1', 25,  24, 73),
       ( 23, '0', 46,  25,  5), ( 24, '1', 31,  25, 73),
       ( 25, '0', 54,  28, 14), ( 26, '1', 29,  28, 82),
       ( 27, '0', 45,  28, 32), ( 28, '1', 35,  28, 61),
       ( 29, '0', 40,  29, 31), ( 30, '0', 23,  29, 87),
       ( 31, '1', 60,  30,  4), ( 32, '0', 21,  30, 73),
       ( 33, '1', 53,  33,  4), ( 34, '1', 18,  33, 92),
       ( 35, '0', 49,  33, 14),

The gender columns still remain as string.

Convert array with dtype void to numpy array.

https://stackoverflow.com/questions/53067064/convert-numpy-void-to-numpy-ndarray

In [144]:
data_list = data.tolist()
data_list

[(1, '1', 19, 15, 39),
 (2, '1', 21, 15, 81),
 (3, '0', 20, 16, 6),
 (4, '0', 23, 16, 77),
 (5, '0', 31, 17, 40),
 (6, '0', 22, 17, 76),
 (7, '0', 35, 18, 6),
 (8, '0', 23, 18, 94),
 (9, '1', 64, 19, 3),
 (10, '0', 30, 19, 72),
 (11, '1', 67, 19, 14),
 (12, '0', 35, 19, 99),
 (13, '0', 58, 20, 15),
 (14, '0', 24, 20, 77),
 (15, '1', 37, 20, 13),
 (16, '1', 22, 20, 79),
 (17, '0', 35, 21, 35),
 (18, '1', 20, 21, 66),
 (19, '1', 52, 23, 29),
 (20, '0', 35, 23, 98),
 (21, '1', 35, 24, 35),
 (22, '1', 25, 24, 73),
 (23, '0', 46, 25, 5),
 (24, '1', 31, 25, 73),
 (25, '0', 54, 28, 14),
 (26, '1', 29, 28, 82),
 (27, '0', 45, 28, 32),
 (28, '1', 35, 28, 61),
 (29, '0', 40, 29, 31),
 (30, '0', 23, 29, 87),
 (31, '1', 60, 30, 4),
 (32, '0', 21, 30, 73),
 (33, '1', 53, 33, 4),
 (34, '1', 18, 33, 92),
 (35, '0', 49, 33, 14),
 (36, '0', 21, 33, 81),
 (37, '0', 42, 34, 17),
 (38, '0', 30, 34, 73),
 (39, '0', 36, 37, 26),
 (40, '0', 20, 37, 75),
 (41, '0', 65, 38, 35),
 (42, '1', 24, 38, 92),
 (43, '

In [145]:
type(data_list)

list

In [146]:
data_new = np.array(data_list)
data_new

array([['1', '1', '19', '15', '39'],
       ['2', '1', '21', '15', '81'],
       ['3', '0', '20', '16', '6'],
       ['4', '0', '23', '16', '77'],
       ['5', '0', '31', '17', '40'],
       ['6', '0', '22', '17', '76'],
       ['7', '0', '35', '18', '6'],
       ['8', '0', '23', '18', '94'],
       ['9', '1', '64', '19', '3'],
       ['10', '0', '30', '19', '72'],
       ['11', '1', '67', '19', '14'],
       ['12', '0', '35', '19', '99'],
       ['13', '0', '58', '20', '15'],
       ['14', '0', '24', '20', '77'],
       ['15', '1', '37', '20', '13'],
       ['16', '1', '22', '20', '79'],
       ['17', '0', '35', '21', '35'],
       ['18', '1', '20', '21', '66'],
       ['19', '1', '52', '23', '29'],
       ['20', '0', '35', '23', '98'],
       ['21', '1', '35', '24', '35'],
       ['22', '1', '25', '24', '73'],
       ['23', '0', '46', '25', '5'],
       ['24', '1', '31', '25', '73'],
       ['25', '0', '54', '28', '14'],
       ['26', '1', '29', '28', '82'],
       ['27', '0', '45', 

All dtype is string. Will convert all to int.

In [147]:
data_new = data_new.astype(int)
data_new

array([[  1,   1,  19,  15,  39],
       [  2,   1,  21,  15,  81],
       [  3,   0,  20,  16,   6],
       [  4,   0,  23,  16,  77],
       [  5,   0,  31,  17,  40],
       [  6,   0,  22,  17,  76],
       [  7,   0,  35,  18,   6],
       [  8,   0,  23,  18,  94],
       [  9,   1,  64,  19,   3],
       [ 10,   0,  30,  19,  72],
       [ 11,   1,  67,  19,  14],
       [ 12,   0,  35,  19,  99],
       [ 13,   0,  58,  20,  15],
       [ 14,   0,  24,  20,  77],
       [ 15,   1,  37,  20,  13],
       [ 16,   1,  22,  20,  79],
       [ 17,   0,  35,  21,  35],
       [ 18,   1,  20,  21,  66],
       [ 19,   1,  52,  23,  29],
       [ 20,   0,  35,  23,  98],
       [ 21,   1,  35,  24,  35],
       [ 22,   1,  25,  24,  73],
       [ 23,   0,  46,  25,   5],
       [ 24,   1,  31,  25,  73],
       [ 25,   0,  54,  28,  14],
       [ 26,   1,  29,  28,  82],
       [ 27,   0,  45,  28,  32],
       [ 28,   1,  35,  28,  61],
       [ 29,   0,  40,  29,  31],
       [ 30,  

In [148]:
data_new.shape

(200, 5)

In [149]:
data_new.ndim

2

### Understand Customer Demographics

In [150]:
header = ['id', 'Gender', 'Age', 'AnnualIncome','Spending']

In [151]:
np.mean(data_new,axis=0)

array([100.5 ,   0.44,  38.85,  60.56,  50.2 ])

We can ignore the first 2 number as it is meaningless for id and gender. The average age is 38.85, average annual income is 60.56k and average spending score is 50.2.

In [152]:
# Extract and filter male customer and female customer
male_customer = data_new[data_new[:,1] == 1]
female_customer = data_new[data_new[:,1] == 0]

In [153]:
# Extract the last 3 columns of age, income and spending score
male_customer  = male_customer[:,2:]
female_customer  = female_customer[:,2:]

### Male Customer Analysis

In [154]:
np.mean(male_customer, axis = 0)

array([39.80681818, 62.22727273, 48.51136364])

For male customer, average age is 39.80, average income is 62.23K and average spending score is 48.5.

In [155]:
np.median(male_customer, axis = 0)

array([37. , 62.5, 50. ])

The median show relatively similar data. This shows that spending distribution is fairly distributed.

In [156]:
np.argmin(male_customer[:,-1])

67

In [157]:
male_customer[np.argmin(male_customer[:,-1])]

array([37, 78,  1])

In [158]:
np.argmax(male_customer[:,-1])

62

In [159]:
male_customer[np.argmax(male_customer[:,-1])]

array([28, 77, 97])

For male customer, minimum spending score is 1, he is a 37 year old with income of 78k. The highest spender is a 28 year old with income of 77k. His spending score is 97.

### Female Customer Analysis

In [160]:
np.mean(female_customer, axis = 0)

array([38.09821429, 59.25      , 51.52678571])

For female customer, average age is 38.1, average income is 59.25K and average spending score is 48.5.

In [161]:
np.median(female_customer, axis = 0)

array([35., 60., 50.])

The median is also very near to the mean, showing even distribution.

In [162]:
female_customer[np.argmin(female_customer[:,-1])]

array([46, 25,  5])

In [163]:
female_customer[np.argmax(female_customer[:,-1])]

array([35, 19, 99])

For female customer, minimum spending score is 5, she is a 46 year old with income of 19k. The highest spender is a 35 year old with income of 19Kk. Her spending score is 99. 

In [164]:
np.sum(male_customer[:,2] > 75)

15

In [165]:
np.sum(female_customer[:,2] > 75)

23

There are more female customer with spending score more than 75 compared to male customer. This shows that female customer has more spending score.

In conclusion, there are more female high spender compare to mal spender. Average income is around 60K which is not high. This indicate that high spender is not necessary with high income and is about. Average age for both male and female customer is about 40 years old. This is also inline with the total average.  


###  Identify High-Value Customers

In [166]:
# Check if filter is correct
data_new[:,-1] > 80

array([False,  True, False, False, False, False, False,  True, False,
       False, False,  True, False, False, False, False, False, False,
       False,  True, False, False, False, False, False,  True, False,
       False, False,  True, False, False, False,  True, False,  True,
       False, False, False, False, False,  True, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False,  True, False, False,
       False,  True,

In [167]:
# Display with filter
high_score = data_new[data_new[:,-1] > 80] 
high_score

array([[  2,   1,  21,  15,  81],
       [  8,   0,  23,  18,  94],
       [ 12,   0,  35,  19,  99],
       [ 20,   0,  35,  23,  98],
       [ 26,   1,  29,  28,  82],
       [ 30,   0,  23,  29,  87],
       [ 34,   1,  18,  33,  92],
       [ 36,   0,  21,  33,  81],
       [ 42,   1,  24,  38,  92],
       [124,   1,  39,  69,  91],
       [128,   1,  40,  71,  95],
       [136,   0,  29,  73,  88],
       [142,   1,  32,  75,  93],
       [144,   0,  32,  76,  87],
       [146,   1,  28,  77,  97],
       [150,   1,  34,  78,  90],
       [152,   1,  39,  78,  88],
       [156,   0,  27,  78,  89],
       [162,   0,  29,  79,  83],
       [164,   0,  31,  81,  93],
       [168,   0,  33,  86,  95],
       [174,   1,  36,  87,  92],
       [176,   0,  30,  88,  86],
       [180,   1,  35,  93,  90],
       [182,   0,  32,  97,  86],
       [184,   0,  29,  98,  88],
       [186,   1,  30,  99,  97],
       [190,   0,  36, 103,  85],
       [194,   0,  38, 113,  91],
       [200,  

In [168]:
np.mean(high_score, axis=0)

array([121.6       ,   0.46666667,  30.6       ,  69.06666667,
        89.76666667])

In [169]:
header

['id', 'Gender', 'Age', 'AnnualIncome', 'Spending']

Average income for the top spender is 69K.

In [170]:
np.argsort(high_score[:,-1])

array([ 0,  7,  4, 18, 29, 27, 24, 22,  5, 13, 25, 16, 11, 17, 15, 23, 28,
        9, 21,  8,  6, 19, 12,  1, 10, 20, 26, 14,  3,  2])

In [171]:
high_score[np.argsort(high_score[:,-1])]

array([[  2,   1,  21,  15,  81],
       [ 36,   0,  21,  33,  81],
       [ 26,   1,  29,  28,  82],
       [162,   0,  29,  79,  83],
       [200,   1,  30, 137,  83],
       [190,   0,  36, 103,  85],
       [182,   0,  32,  97,  86],
       [176,   0,  30,  88,  86],
       [ 30,   0,  23,  29,  87],
       [144,   0,  32,  76,  87],
       [184,   0,  29,  98,  88],
       [152,   1,  39,  78,  88],
       [136,   0,  29,  73,  88],
       [156,   0,  27,  78,  89],
       [150,   1,  34,  78,  90],
       [180,   1,  35,  93,  90],
       [194,   0,  38, 113,  91],
       [124,   1,  39,  69,  91],
       [174,   1,  36,  87,  92],
       [ 42,   1,  24,  38,  92],
       [ 34,   1,  18,  33,  92],
       [164,   0,  31,  81,  93],
       [142,   1,  32,  75,  93],
       [  8,   0,  23,  18,  94],
       [128,   1,  40,  71,  95],
       [168,   0,  33,  86,  95],
       [186,   1,  30,  99,  97],
       [146,   1,  28,  77,  97],
       [ 20,   0,  35,  23,  98],
       [ 12,  

In [172]:
sorted_high_score = high_score[np.argsort(high_score[:,-1])][::-1]
sorted_high_score

array([[ 12,   0,  35,  19,  99],
       [ 20,   0,  35,  23,  98],
       [146,   1,  28,  77,  97],
       [186,   1,  30,  99,  97],
       [168,   0,  33,  86,  95],
       [128,   1,  40,  71,  95],
       [  8,   0,  23,  18,  94],
       [142,   1,  32,  75,  93],
       [164,   0,  31,  81,  93],
       [ 34,   1,  18,  33,  92],
       [ 42,   1,  24,  38,  92],
       [174,   1,  36,  87,  92],
       [124,   1,  39,  69,  91],
       [194,   0,  38, 113,  91],
       [180,   1,  35,  93,  90],
       [150,   1,  34,  78,  90],
       [156,   0,  27,  78,  89],
       [136,   0,  29,  73,  88],
       [152,   1,  39,  78,  88],
       [184,   0,  29,  98,  88],
       [144,   0,  32,  76,  87],
       [ 30,   0,  23,  29,  87],
       [176,   0,  30,  88,  86],
       [182,   0,  32,  97,  86],
       [190,   0,  36, 103,  85],
       [200,   1,  30, 137,  83],
       [162,   0,  29,  79,  83],
       [ 26,   1,  29,  28,  82],
       [ 36,   0,  21,  33,  81],
       [  2,  

- https://stackoverflow.com/questions/2828059/sorting-arrays-in-numpy-by-column
- https://stackoverflow.com/questions/26984414/efficiently-sorting-a-numpy-array-in-descending-order

In [173]:
top_10 = sorted_high_score[:10,:]
top_10 

array([[ 12,   0,  35,  19,  99],
       [ 20,   0,  35,  23,  98],
       [146,   1,  28,  77,  97],
       [186,   1,  30,  99,  97],
       [168,   0,  33,  86,  95],
       [128,   1,  40,  71,  95],
       [  8,   0,  23,  18,  94],
       [142,   1,  32,  75,  93],
       [164,   0,  31,  81,  93],
       [ 34,   1,  18,  33,  92]])

In [174]:
np.mean(top_10, axis = 0)

array([100.8,   0.5,  30.5,  58.2,  95.3])

The average income of the top 10 customer is only 58K. Average age is only 31 years old. The top spender is a 35 year old female with income of only 19K. The number of male and female customer is even. The oldest top spender is a 40 year old male with income of 71K.

In conclusion, the top spenders average income is inline with total average income. The average age of the top customer is lower then the total average, 

### Explore Relationships Between Features


In [175]:
data_new[:10]

array([[ 1,  1, 19, 15, 39],
       [ 2,  1, 21, 15, 81],
       [ 3,  0, 20, 16,  6],
       [ 4,  0, 23, 16, 77],
       [ 5,  0, 31, 17, 40],
       [ 6,  0, 22, 17, 76],
       [ 7,  0, 35, 18,  6],
       [ 8,  0, 23, 18, 94],
       [ 9,  1, 64, 19,  3],
       [10,  0, 30, 19, 72]])

In [176]:
header

['id', 'Gender', 'Age', 'AnnualIncome', 'Spending']

In [177]:
age = data_new[:,2:3].ravel()
age

array([19, 21, 20, 23, 31, 22, 35, 23, 64, 30, 67, 35, 58, 24, 37, 22, 35,
       20, 52, 35, 35, 25, 46, 31, 54, 29, 45, 35, 40, 23, 60, 21, 53, 18,
       49, 21, 42, 30, 36, 20, 65, 24, 48, 31, 49, 24, 50, 27, 29, 31, 49,
       33, 31, 59, 50, 47, 51, 69, 27, 53, 70, 19, 67, 54, 63, 18, 43, 68,
       19, 32, 70, 47, 60, 60, 59, 26, 45, 40, 23, 49, 57, 38, 67, 46, 21,
       48, 55, 22, 34, 50, 68, 18, 48, 40, 32, 24, 47, 27, 48, 20, 23, 49,
       67, 26, 49, 21, 66, 54, 68, 66, 65, 19, 38, 19, 18, 19, 63, 49, 51,
       50, 27, 38, 40, 39, 23, 31, 43, 40, 59, 38, 47, 39, 25, 31, 20, 29,
       44, 32, 19, 35, 57, 32, 28, 32, 25, 28, 48, 32, 34, 34, 43, 39, 44,
       38, 47, 27, 37, 30, 34, 30, 56, 29, 19, 31, 50, 36, 42, 33, 36, 32,
       40, 28, 36, 36, 52, 30, 58, 27, 59, 35, 37, 32, 46, 29, 41, 30, 54,
       28, 41, 36, 34, 32, 33, 38, 47, 35, 45, 32, 32, 30])

In [178]:
spending = data_new[:,-1]
spending

array([39, 81,  6, 77, 40, 76,  6, 94,  3, 72, 14, 99, 15, 77, 13, 79, 35,
       66, 29, 98, 35, 73,  5, 73, 14, 82, 32, 61, 31, 87,  4, 73,  4, 92,
       14, 81, 17, 73, 26, 75, 35, 92, 36, 61, 28, 65, 55, 47, 42, 42, 52,
       60, 54, 60, 45, 41, 50, 46, 51, 46, 56, 55, 52, 59, 51, 59, 50, 48,
       59, 47, 55, 42, 49, 56, 47, 54, 53, 48, 52, 42, 51, 55, 41, 44, 57,
       46, 58, 55, 60, 46, 55, 41, 49, 40, 42, 52, 47, 50, 42, 49, 41, 48,
       59, 55, 56, 42, 50, 46, 43, 48, 52, 54, 42, 46, 48, 50, 43, 59, 43,
       57, 56, 40, 58, 91, 29, 77, 35, 95, 11, 75,  9, 75, 34, 71,  5, 88,
        7, 73, 10, 72,  5, 93, 40, 87, 12, 97, 36, 74, 22, 90, 17, 88, 20,
       76, 16, 89,  1, 78,  1, 73, 35, 83,  5, 93, 26, 75, 20, 95, 27, 63,
       13, 75, 10, 92, 13, 86, 15, 69, 14, 90, 32, 86, 15, 88, 39, 97, 24,
       68, 17, 85, 23, 69,  8, 91, 16, 79, 28, 74, 18, 83])

In [179]:
income = data_new[:,-2]
income

array([ 15,  15,  16,  16,  17,  17,  18,  18,  19,  19,  19,  19,  20,
        20,  20,  20,  21,  21,  23,  23,  24,  24,  25,  25,  28,  28,
        28,  28,  29,  29,  30,  30,  33,  33,  33,  33,  34,  34,  37,
        37,  38,  38,  39,  39,  39,  39,  40,  40,  40,  40,  42,  42,
        43,  43,  43,  43,  44,  44,  46,  46,  46,  46,  47,  47,  48,
        48,  48,  48,  48,  48,  49,  49,  50,  50,  54,  54,  54,  54,
        54,  54,  54,  54,  54,  54,  54,  54,  57,  57,  58,  58,  59,
        59,  60,  60,  60,  60,  60,  60,  61,  61,  62,  62,  62,  62,
        62,  62,  63,  63,  63,  63,  63,  63,  64,  64,  65,  65,  65,
        65,  67,  67,  67,  67,  69,  69,  70,  70,  71,  71,  71,  71,
        71,  71,  72,  72,  73,  73,  73,  73,  74,  74,  75,  75,  76,
        76,  77,  77,  77,  77,  78,  78,  78,  78,  78,  78,  78,  78,
        78,  78,  78,  78,  79,  79,  81,  81,  85,  85,  86,  86,  87,
        87,  87,  87,  87,  87,  88,  88,  88,  88,  93,  93,  9

In [180]:
age_spending = np.corrcoef(age,spending)
age_spending

array([[ 1.        , -0.32722685],
       [-0.32722685,  1.        ]])

This shows that there is an inverse relationship between age and spending. Our previous analysis also shows that the top spender are not the oldest.

In [181]:
income_spending = np.corrcoef(income, spending)
income_spending

array([[1.        , 0.00990285],
       [0.00990285, 1.        ]])

This shows that there is very little correlation between spending score and income.

In [182]:
age_income = np.corrcoef(age,income)
age_income

array([[ 1.        , -0.01239804],
       [-0.01239804,  1.        ]])

This shows that there is very little correlation between age and income.

In [183]:
# Filter young adults
(data_new[:,2] > 18) & (data_new[:,2] < 25)

array([ True,  True,  True,  True, False,  True, False,  True, False,
       False, False, False, False,  True, False,  True, False,  True,
       False, False, False, False, False, False, False, False, False,
       False, False,  True, False,  True, False, False, False,  True,
       False, False, False,  True, False,  True, False, False, False,
        True, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False,  True, False,
       False, False, False, False, False,  True, False, False, False,
       False, False, False, False, False, False,  True, False, False,
       False, False, False,  True, False, False,  True, False, False,
       False, False, False, False, False,  True, False, False, False,
        True,  True, False, False, False, False,  True, False, False,
       False, False, False,  True, False,  True, False,  True, False,
       False, False, False, False, False, False, False,  True, False,
       False, False,

In [184]:
data_new[(data_new[:,2] > 18) & (data_new[:,2] < 25)]

array([[  1,   1,  19,  15,  39],
       [  2,   1,  21,  15,  81],
       [  3,   0,  20,  16,   6],
       [  4,   0,  23,  16,  77],
       [  6,   0,  22,  17,  76],
       [  8,   0,  23,  18,  94],
       [ 14,   0,  24,  20,  77],
       [ 16,   1,  22,  20,  79],
       [ 18,   1,  20,  21,  66],
       [ 30,   0,  23,  29,  87],
       [ 32,   0,  21,  30,  73],
       [ 36,   0,  21,  33,  81],
       [ 40,   0,  20,  37,  75],
       [ 42,   1,  24,  38,  92],
       [ 46,   0,  24,  39,  65],
       [ 62,   1,  19,  46,  55],
       [ 69,   1,  19,  48,  59],
       [ 79,   0,  23,  54,  52],
       [ 85,   0,  21,  54,  57],
       [ 88,   0,  22,  57,  55],
       [ 96,   1,  24,  60,  52],
       [100,   1,  20,  61,  49],
       [101,   0,  23,  62,  41],
       [106,   0,  21,  62,  42],
       [112,   0,  19,  63,  54],
       [114,   1,  19,  64,  46],
       [116,   0,  19,  65,  50],
       [125,   0,  23,  70,  29],
       [135,   1,  20,  73,   5],
       [139,  

In [185]:
young_adults = data_new[(data_new[:,2] > 18) & (data_new[:,2] < 25)]
young_adults

array([[  1,   1,  19,  15,  39],
       [  2,   1,  21,  15,  81],
       [  3,   0,  20,  16,   6],
       [  4,   0,  23,  16,  77],
       [  6,   0,  22,  17,  76],
       [  8,   0,  23,  18,  94],
       [ 14,   0,  24,  20,  77],
       [ 16,   1,  22,  20,  79],
       [ 18,   1,  20,  21,  66],
       [ 30,   0,  23,  29,  87],
       [ 32,   0,  21,  30,  73],
       [ 36,   0,  21,  33,  81],
       [ 40,   0,  20,  37,  75],
       [ 42,   1,  24,  38,  92],
       [ 46,   0,  24,  39,  65],
       [ 62,   1,  19,  46,  55],
       [ 69,   1,  19,  48,  59],
       [ 79,   0,  23,  54,  52],
       [ 85,   0,  21,  54,  57],
       [ 88,   0,  22,  57,  55],
       [ 96,   1,  24,  60,  52],
       [100,   1,  20,  61,  49],
       [101,   0,  23,  62,  41],
       [106,   0,  21,  62,  42],
       [112,   0,  19,  63,  54],
       [114,   1,  19,  64,  46],
       [116,   0,  19,  65,  50],
       [125,   0,  23,  70,  29],
       [135,   1,  20,  73,   5],
       [139,  

In [186]:
young_adults.mean(axis=0)

array([64.12903226,  0.41935484, 21.19354839, 43.80645161, 55.77419355])

In [187]:
older_adults = data_new[data_new[:,2] > 25]

In [188]:
older_adults.mean(axis = 0)

array([108.05555556,   0.43209877,  43.        ,  64.04938272,
        49.08641975])

Average income of young adults is 44K with average spending score of 55.77. Adults older than 25 has average income of 64K with spending score of 49. This shows that young adults has high spending score although the income is lower.  

### Customer Segmentation


In [189]:
header

['id', 'Gender', 'Age', 'AnnualIncome', 'Spending']

In [190]:
data_new[(data_new[:,-1] >= 1) & (data_new[:,-1] <= 33)]

array([[  3,   0,  20,  16,   6],
       [  7,   0,  35,  18,   6],
       [  9,   1,  64,  19,   3],
       [ 11,   1,  67,  19,  14],
       [ 13,   0,  58,  20,  15],
       [ 15,   1,  37,  20,  13],
       [ 19,   1,  52,  23,  29],
       [ 23,   0,  46,  25,   5],
       [ 25,   0,  54,  28,  14],
       [ 27,   0,  45,  28,  32],
       [ 29,   0,  40,  29,  31],
       [ 31,   1,  60,  30,   4],
       [ 33,   1,  53,  33,   4],
       [ 35,   0,  49,  33,  14],
       [ 37,   0,  42,  34,  17],
       [ 39,   0,  36,  37,  26],
       [ 45,   0,  49,  39,  28],
       [125,   0,  23,  70,  29],
       [129,   1,  59,  71,  11],
       [131,   1,  47,  71,   9],
       [135,   1,  20,  73,   5],
       [137,   0,  44,  73,   7],
       [139,   1,  19,  74,  10],
       [141,   0,  57,  75,   5],
       [145,   1,  25,  77,  12],
       [149,   0,  34,  78,  22],
       [151,   1,  43,  78,  17],
       [153,   0,  44,  78,  20],
       [155,   0,  47,  78,  16],
       [157,  

In [191]:
low = data_new[(data_new[:,-1] >= 1) & (data_new[:,-1] <= 33)]

In [192]:
data_new[(data_new[:,-1] >= 34) & (data_new[:,-1] <= 66)]

array([[  1,   1,  19,  15,  39],
       [  5,   0,  31,  17,  40],
       [ 17,   0,  35,  21,  35],
       [ 18,   1,  20,  21,  66],
       [ 21,   1,  35,  24,  35],
       [ 28,   1,  35,  28,  61],
       [ 41,   0,  65,  38,  35],
       [ 43,   1,  48,  39,  36],
       [ 44,   0,  31,  39,  61],
       [ 46,   0,  24,  39,  65],
       [ 47,   0,  50,  40,  55],
       [ 48,   0,  27,  40,  47],
       [ 49,   0,  29,  40,  42],
       [ 50,   0,  31,  40,  42],
       [ 51,   0,  49,  42,  52],
       [ 52,   1,  33,  42,  60],
       [ 53,   0,  31,  43,  54],
       [ 54,   1,  59,  43,  60],
       [ 55,   0,  50,  43,  45],
       [ 56,   1,  47,  43,  41],
       [ 57,   0,  51,  44,  50],
       [ 58,   1,  69,  44,  46],
       [ 59,   0,  27,  46,  51],
       [ 60,   1,  53,  46,  46],
       [ 61,   1,  70,  46,  56],
       [ 62,   1,  19,  46,  55],
       [ 63,   0,  67,  47,  52],
       [ 64,   0,  54,  47,  59],
       [ 65,   1,  63,  48,  51],
       [ 66,  

In [193]:
medium = data_new[(data_new[:,-1] >= 34) & (data_new[:,-1] <= 66)]

In [194]:
high = data_new[(data_new[:,-1] >= 67) & (data_new[:,-1] <= 100)]

In [195]:
header

['id', 'Gender', 'Age', 'AnnualIncome', 'Spending']

In [196]:
low.mean(axis=0)

array([115.53061224,   0.48979592,  42.87755102,  67.        ,
        15.30612245])

For the group with lower spending score, the average age is 43 years old with average annual income of 67K. 

In [197]:
medium.mean(axis=0)

array([83.77659574,  0.42553191, 42.0106383 , 53.86170213, 48.89361702])

For the group with medium spending score, the average age is 42 years old with average annual income of 53K. 

In [198]:
high.mean(axis=0)

array([115.15789474,   0.42105263,  30.1754386 ,  66.07017544,
        82.35087719])

For the group with high spending score, the average age is 30 years old with average annual income of 66K. 

In conclusion, the high spending group belongs to the younger working adults specially the group 40 years old and below. Their income is slightly above average.

### Find Behavioral Similarities


In [199]:
data_new[:10]

array([[ 1,  1, 19, 15, 39],
       [ 2,  1, 21, 15, 81],
       [ 3,  0, 20, 16,  6],
       [ 4,  0, 23, 16, 77],
       [ 5,  0, 31, 17, 40],
       [ 6,  0, 22, 17, 76],
       [ 7,  0, 35, 18,  6],
       [ 8,  0, 23, 18, 94],
       [ 9,  1, 64, 19,  3],
       [10,  0, 30, 19, 72]])

In [200]:
header

['id', 'Gender', 'Age', 'AnnualIncome', 'Spending']

In [201]:
avg = np.mean(data_new, axis=0)

In [202]:
std_dev = np.std(data_new, axis=0)

In [203]:
data_norm = (data_new - avg) / std_dev
data_norm

array([[-1.7234121 ,  1.12815215, -1.42456879, -1.73899919, -0.43480148],
       [-1.70609137,  1.12815215, -1.28103541, -1.73899919,  1.19570407],
       [-1.68877065, -0.88640526, -1.3528021 , -1.70082976, -1.71591298],
       [-1.67144992, -0.88640526, -1.13750203, -1.70082976,  1.04041783],
       [-1.6541292 , -0.88640526, -0.56336851, -1.66266033, -0.39597992],
       [-1.63680847, -0.88640526, -1.20926872, -1.66266033,  1.00159627],
       [-1.61948775, -0.88640526, -0.27630176, -1.62449091, -1.71591298],
       [-1.60216702, -0.88640526, -1.13750203, -1.62449091,  1.70038436],
       [-1.5848463 ,  1.12815215,  1.80493225, -1.58632148, -1.83237767],
       [-1.56752558, -0.88640526, -0.6351352 , -1.58632148,  0.84631002],
       [-1.55020485,  1.12815215,  2.02023231, -1.58632148, -1.4053405 ],
       [-1.53288413, -0.88640526, -0.27630176, -1.58632148,  1.89449216],
       [-1.5155634 , -0.88640526,  1.37433211, -1.54815205, -1.36651894],
       [-1.49824268, -0.88640526, -1.0

In [204]:
data_norm = data_norm[:, 2:]

In [205]:
data_norm.shape

(200, 3)

- https://www.geeksforgeeks.org/calculate-the-euclidean-distance-using-numpy/
- https://stackoverflow.com/questions/1401712/how-can-the-euclidean-distance-be-calculated-with-numpy
- https://www.kdnuggets.com/2023/03/distance-metrics-euclidean-manhattan-minkowski-oh.html

In [206]:
highest_idx = np.argmax(data_norm[:,-1]) 
highest_spender = data_norm[highest_idx]
highest_spender

array([-0.27630176, -1.58632148,  1.89449216])

In [207]:
distances = np.linalg.norm(data_norm - highest_spender, axis=1)
distances.shape

(200,)

In [208]:
distances

array([2.60142971, 1.23333086, 3.76921616, 1.21828611, 2.30965307,
       1.29364526, 3.61060691, 0.88362925, 4.2686173 , 1.10790218,
       4.02031893, 0.        , 3.65516654, 1.16366029, 3.34195613,
       1.21438449, 2.48575237, 1.67509024, 2.98272515, 0.15753602,
       2.49189884, 1.25310701, 3.7406563 , 1.07408784, 3.58695137,
       0.85964041, 2.72001617, 1.51468853, 2.69134635, 1.05089524,
       4.12274443, 1.48478417, 3.94410976, 1.35936908, 3.49054992,
       1.33542052, 3.2732259 , 1.21464971, 2.91694978, 1.58081869,
       3.36667373, 1.10589185, 2.72670557, 1.68565757, 3.03143718,
       1.71702965, 2.17235359, 2.24663296, 2.3925977 , 2.37097334,
       2.26039964, 1.7560252 , 1.99336143, 2.46944546, 2.52839349,
       2.57890935, 2.4181941 , 3.33136264, 2.20547114, 2.63899874,
       3.18716923, 2.30181951, 3.12177776, 2.3265663 , 2.95560585,
       2.26387309, 2.27452553, 3.27934571, 2.22601803, 2.31232494,
       3.24627439, 2.63618905, 2.89601555, 2.72135538, 2.97096

In [209]:
similar_customers_idx = distances <= 1.0
similar_customers = data_new[similar_customers_idx]
similar_customers

array([[ 8,  0, 23, 18, 94],
       [12,  0, 35, 19, 99],
       [20,  0, 35, 23, 98],
       [26,  1, 29, 28, 82]])

Younger female customer has similar high spending score.

The following are code from ChatGPT

- https://chatgpt.com/share/67988b3a-8488-8000-bed0-af7edf4d92e2

### EDA Findings

1. There are more female high spender compare to male spender. 
2. Average income is around 60K which is not high. This indicate that high spender is not necessary with high income and is about. Average age for both male and female customer is about 40 years old. This is also inline with the total average.  
3. The average income of the top 10 customer is only 58K. Average age is only 31 years old. The top spender is a 35 year old female with income of only 19K. The number of male and female customer is even.
4. The top spenders average income is inline with total average income. The average age of the top customer is lower then the total average, 
5. There is little or no correlation between income and spending score. Although there is a negative correlation between age and spending. This indicate that the younger adults are the high spender.
6. Average income of young adults is 44K with average spending score of 55.77. Adults older than 25 has average income of 64K with spending score of 49. This shows that young adults has high spending score although the income is lower.  
7. The high spending group belongs to the younger working adults specially the group 40 years old and below. Their income is slightly above average.
8. Young adult female customers has similar score with the top spender.


Based on the findings above, the mall management should concentrate on younger working adults between 25 to 40 years old.